### PR TITLE
Move version check earlier

### LINKED ISSUE #6659: RAR Cache should be versioned (TranslateDictionary)
I happened to be building under debugger, and saw the serialization exception in DateTime ctor (which usually signals corrupt data) when reading the state file: `obj\Debug\TestSolution.csproj.AssemblyReference.cache`

The .cache file is attached:
[TestSolution.csproj.AssemblyReference.zip](https://github.com/dotnet/msbuild/files/6787793/TestSolution.csproj.AssemblyReference.zip)

```
mscorlib	DateTime..ctor Line 159
Microsoft.Build.Tasks.Core	BinaryTranslator.BinaryReadTranslator.Translate Line 330
Microsoft.Build.Tasks.Core	SystemState.FileState.Translate Line 164
Microsoft.Build.Tasks.Core	SystemState.FileState..ctor Line 154
Microsoft.Build.Tasks.Core	SystemState.<>c.<Translate>b__18_0 Line 251
Microsoft.Build.Tasks.Core	TranslatorHelpers.Translate Line 38
Microsoft.Build.Tasks.Core	TranslatorHelpers.<>c__DisplayClass1_0`1.<AdaptFactory>g__TranslateUsingValueFactory|0 Line 51
Microsoft.Build.Tasks.Core	BinaryTranslator.BinaryReadTranslator.TranslateDictionary Line 600
Microsoft.Build.Tasks.Core	TranslatorHelpers.TranslateDictionary Line 88
Microsoft.Build.Tasks.Core	SystemState.Translate Line 255
Microsoft.Build.Tasks.Core	SystemState..ctor Line 223
	[Internal Frame, 'U-->M']
mscorlib	RuntimeConstructorInfo.Invoke Line 753
mscorlib	ConstructorInfo.Invoke Line 105
Microsoft.Build.Tasks.Core	StateFileBase.DeserializeCache Line 88
Microsoft.Build.Tasks.Core	ResolveAssemblyReference.ReadStateFile Line 2006
Microsoft.Build.Tasks.Core	ResolveAssemblyReference.Execute Line 2262
Microsoft.Build.Tasks.Core	ResolveAssemblyReference.Execute Line 3150
```

Since there's a try catch around reading the cache, the error is benign (I think?).

But I'd be curious what happened, either the format of serialized dictionaries changed across MSBuild versions or something else.

@Forgind any ideas?

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #6659

The version check was previously after having deserialized everything else. This doesn't make sense and can lead to errors. In this PR, I moved it up.

### CODE REVIEW

File: src/Tasks/StateFileBase.cs [hunk 1]
```diff
@@ -77,8 +77,17 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
...
                     {
                         using var translator = BinaryTranslator.GetReadTranslator(s, buffer: null);
+
                         byte version = 0;
                         translator.Translate(ref version);
+                        // If retVal is still null or the version is wrong, log a message not a warning. This could be a valid cache with the wrong version preventing correct deserialization.
+                        // For the latter case, internals may be unexpectedly null.
+                        if (retVal == null || version != CurrentSerializationVersion)
+                        {
+                            log.LogMessageFromResources("General.CouldNotReadStateFileMessage", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
```
Comment (rainersigwald) [inline src/Tasks/StateFileBase.cs:87]: I think it continues to make sense for this to be a quiet message. But do we have a test in place that this is _not_ the case for freshly-built-everything using the SDK cache?

**Review by rainersigwald — APPROVED** [review]

File: src/Tasks/StateFileBase.cs [hunk 2]
```diff
@@ -77,8 +77,17 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
                     using (FileStream s = File.OpenRead(stateFile))
                     {
                         using var translator = BinaryTranslator.GetReadTranslator(s, buffer: null);
+
                         byte version = 0;
                         translator.Translate(ref version);
+                        // If retVal is still null or the version is wrong, log a message not a warning. This could be a valid cache with the wrong version preventing correct deserialization.
+                        // For the latter case, internals may be unexpectedly null.
+                        if (retVal == null || version != CurrentSerializationVersion)
```
Comment (rainersigwald) [inline src/Tasks/StateFileBase.cs:85]: retval will always be null here, right?
```suggestion
                        if (version != CurrentSerializationVersion)
```

File: src/Tasks/StateFileBase.cs (hunk 1 above)
Comment (Forgind) [reply src/Tasks/StateFileBase.cs:87]: We do not. If I remember correctly, we decided it would be too hard for some reason.

Comment (rainersigwald) [reply src/Tasks/StateFileBase.cs:87]: Can we at least get it added to the manual SDK tests?

File: src/Tasks/StateFileBase.cs [hunk 3]
```diff
@@ -77,8 +77,17 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
                     using (FileStream s = File.OpenRead(stateFile))
                     {
                         using var translator = BinaryTranslator.GetReadTranslator(s, buffer: null);
+
                         byte version = 0;
                         translator.Translate(ref version);
+                        // If retVal is still null or the version is wrong, log a message not a warning. This could be a valid cache with the wrong version preventing correct deserialization.
+                        // For the latter case, internals may be unexpectedly null.
```
Comment (rainersigwald) [inline src/Tasks/StateFileBase.cs:84]: Update comment

File: src/Tasks/StateFileBase.cs (hunk 1 above)
Comment (Forgind) [reply src/Tasks/StateFileBase.cs:87]: Yeah, that sounds like a good idea to me. @marcpopMSFT, agree? If so, I can write up some (very detailed) instructions on testing it.

Comment (marcpopMSFT) [reply src/Tasks/StateFileBase.cs:87]: Please reach out to Richa with the scenario you want covered.  The difficulty with fresh machine tests as you can only have one.

**Review by BenVillalobos — APPROVED** [review]
LGTM

> Please reach out to Richa with the scenario you want covered. The difficulty with fresh machine tests as you can only have one.

What was the status on this? A manual SDK test was mentioned

**Comment by Forgind** [comment]
> LGTM
> 
> > Please reach out to Richa with the scenario you want covered. The difficulty with fresh machine tests as you can only have one.
> 
> What was the status on this? A manual SDK test was mentioned

I still need to write instructions. Bit behind on that front.

**Pull request merged**